pkg/compiler/compilation_document.go

```diff
@@ -59,3 +64,86 @@ func (doc *CompilationDocument) OutputTo(dest string) error {
 	}
 	return nil
 }
+
+func (document *CompilationDocument) OutputResources() (resourceCounts map[string]int, err error) {
+	outDir := document.Configuration.OutDir
+	result := document.Constructs
+
+	err = os.MkdirAll(outDir, 0777)
+	if err != nil {
+		return
+	}
+
+	resourceCounts = make(map[string]int)
+	var resourcesOutput []interface{}
+	var merr multierr.Error
+	for _, construct := range result.ListConstructs() {
+		resourceCounts[construct.Provenance().Capability] = resourceCounts[construct.Provenance().Capability] + 1
+
+		switch r := construct.(type) {
+		case *core.ExecutionUnit:
+			resOut := map[string]interface{}{
+				"Type": r.Provenance().Capability,
+				"Name": r.Provenance().ID,
+			}
+			var files []string
+			for _, f := range r.Files() {
+				files = append(files, f.Path())
+			}
+			resOut["Files"] = files
+			resourcesOutput = append(resourcesOutput, resOut)
+		default:
+			resourcesOutput = append(resourcesOutput, r)
+		}
+
+		output, ok := construct.(core.HasLocalOutput)
+		if !ok {
+			continue
+		}
+		zap.L().Debug("Output", zap.String("type", construct.Provenance().Capability), zap.String("name", construct.Provenance().ID))
+		err = output.OutputTo(outDir)
+		if err != nil {
+			merr.Append(errors.Wrapf(err, "error outputting resource %+v", construct.Provenance()))
+		}
+	}
+
+	f, err := os.Create(path.Join(outDir, "resources.json"))
+	if err != nil {
+		merr.Append(errors.Wrap(err, "error creating resource dump"))
+	} else {
+		defer f.Close()
+
+		enc := json.NewEncoder(f)
+		enc.SetIndent("", "  ")
+		err = enc.Encode(resourcesOutput)
+		if err != nil {
+			merr.Append(errors.Wrap(err, "error writing resources"))
+		}
+	}
+
+	merr.Append(document.OutputHelpers(outDir))
+
+	merr.Append(document.OutputTo(document.Configuration.OutDir))
+	err = merr.ErrOrNil()
+
+	return
+}
+
+func (document *CompilationDocument) OutputHelpers(outDir string) error {
```

Comment:
👀 note to self (& other reviewers if interested): this is new in this copy of the code, relative to the `-` above)

```diff
--- 	func OutputHelpers(dag *core.ResourceGraph, outDir string) error {
+++	func (document *CompilationDocument) OutputHelpers(outDir string) error {
```